nixos/modules/services/network-filesystems/samba.nix

```diff
@@ -214,12 +214,9 @@ in
             }
           ];
         # Always provide a smb.conf to shut up programs like smbclient and smbspool.
-        environment.etc = singleton
-          { source =
-              if cfg.enable then configFile
-              else pkgs.writeText "smb-dummy.conf" "# Samba is disabled.";
-            target = "samba/smb.conf";
-          };
+        environment.etc."samba/smb.conf".source =
+          if cfg.enable then configFile
+          else mkOptionDefault (pkgs.writeText "smb-dummy.conf" "# Samba is disabled.");
```

Comment:
Use `mkDefault` instead

Reply:
With `mkDefault` I can't get `mkForce` to override it.

Reply:
See #31860 for reference.

Reply:
Are you sure? From their definition this doesn't make any sense: https://github.com/NixOS/nixpkgs/blob/fcb425427688c4849e8cbb32582d124b79b9bb44/lib/modules.nix#L496-L498

Reply:
Yes, I had a look at this bit of code as well and can only assume that `mkOverride` is doing magic. I've just tried it again and get the following error:

    The unique option `environment.etc.samba/smb.conf.source' is defined multiple times, in …

Reply:
@jfrankenau What is your `environment.etc."samba/smb.conf"` configuration when you get that error? Maybe you are using the `text` attribute?

Using `mkOptionDefault` here is necessary because the `etc.nix` module uses `mkDefault` to set the `source` field when the `text` field is provided:
https://github.com/NixOS/nixpkgs/blob/58384b1dd0363964ca4a85325b8609ecaca8c91e/nixos/modules/system/etc/etc.nix#L135-L137

That is, if you in this patch use `mkDefault`, have `cfg.enable == false`, and you set
```
environment.etc."samba/smb.conf".text = "blah";
```
then both this code and the code in `etc.nix` will attempt to set the option value at the `mkDefault` priority level.

With `mkOptionDefault` the priorities work out better, if you don't provide a text then the code in `etc.nix` will not trigger and your `source` field will remain intact. The used can also override `source` easily (without using `mkForce`) since the default priority is greater than the option default one. The user can also override the file content using the `text` field since the default priority (set by `etc.nix`) will override the option default priority in this file.

Admittedly, it's not entirely straight-forward but works out for the best in the end 🙂

Reply:
In other words, as far as I can tell this PR looks good.

Reply:
Oh I see, but shouldn't the best course of action then be to change the default priority of the `source` value? Something like `mkOverride 1200`. Because `mkOptionDefault` is indicated to be the options default (aka the thing you get for `options.foo.default`).

Reply:
@rycee, yes I am using the `text` attribute. Thank you for your thorough explanation.

@Infinisil, I agree that `mkOptionDefault` is semantically incorrect. However if at some point the definition for the named mkOverrides ever changes the hardcoded values will have to be updated as well.

Reply:
All right, one more thing though: I think `mkOptionDefault` should include the if statement, such that you can override it the same way if samba is enabled.

Reply:
Done.